crates/pathfinder/src/sequencer.rs

```diff
@@ -37,17 +47,21 @@ impl Client {
     }
 
     /// Gets block by number.
-    pub async fn block_by_number(&self, block_number: BlockNumberOrTag) -> Result<reply::Block> {
+    pub async fn block_by_number(
+        &self,
+        block_number: BlockNumberOrTag,
+    ) -> Result<reply::Block, SequencerError> {
         let block_hash = match block_number {
             BlockNumberOrTag::Number(n) => {
                 let resp = reqwest::get(
                     self.build_query("get_block_hash_by_id", &[("blockId", &n.to_string())]),
                 )
                 .await
                 .unwrap();
+                check_for_error(&resp)?;
                 let resp = resp.text().await?;
 
-                if let Ok(e) = serde_json::from_str::<self::reply::starknet::Error>(resp.as_str()) {
+                if let Ok(e) = serde_json::from_str::<StarknetError>(resp.as_str()) {
```

Comment:
Why does this function have this error check, but many (all?) the others don't?

Reply:
This is quite an interesting example :smile_cat: of how the sequencer can reply:
* the valid answer is not a valid JSON but just a quoted string, for example: `"0x1234"`
* but if starknet reports an error then we have a valid JSON reply, for example: `{code: "StarknetErrorCode.BLOCK_NOT_FOUND", message:"Some error message"}`

So what the function does is:
1. HTTP GET to try and figure out a hash for a given block number
2. if the status code of the reply is an error and other than 500 - create an appropriate `reqwest::Error` which upon `?` operator is transformed into `SequencerError::Transport`
3. Attempt to get the text of the body of the reply
4. It could be a JSON with a starknet error, so attempt to deserialize
5. If it indeed is an error - then return it, it will get wrapped in `SequencerError::StarknetError`
6. Attempt to remove the surplus `"` and parse the remaining unquoted literal into a hash value; if there's a failure at any point - return a `SequencerError::ParseError`
7. Return a valid hash value if all of the above problems were avoided

Most of the other queries reply with a valid JSON, hence the special handling is not required :shrug:

Reply:
> valid JSON but just a quoted string, for example: "0x1234"

to be pedantic, a plain quoted string is valid JSON. it might not be expected json but it's still valid.

Reply:
Yes, you are right :facepalm: , let me re-try with `serde_json` then, as what happens here looks like an obvious overkill.

[Pages 4 & 5 of the rfc](https://www.ietf.org/rfc/rfc7159.txt):
```
   A JSON text is a serialized value.  Note that certain previous
   specifications of JSON constrained a JSON text to be an object or an
   array. Implementations that generate only objects or arrays where a
   JSON text is called for will be interoperable in the sense that all
   implementations will accept these as conforming JSON texts.
```